nixos/modules/programs/java.nix

```
@@ -40,12 +39,35 @@ in
         type = types.package;
       };
 
+      binfmt = mkEnableOption (lib.mdDoc "binfmt to execute java jar's and classes");
+
     };
 
   };
 
   config = mkIf cfg.enable {
 
+    boot.binfmt.registrations = mkIf cfg.binfmt {
+      java-class = {
+        recognitionType = "extension";
+        magicOrExtension = "class";
+        interpreter = pkgs.writeShellScript "java-class-wrapper" ''
+          test -e ${cfg.package}/nix-support/setup-hook && source ${cfg.package}/nix-support/setup-hook
+          classpath=$(dirname "$1")
+          class=$(basename "''${1%%.class}")
+          $JAVA_PATH -classpath "$classpath" "$class" "''${@:2}"
```

```suggestion
          $JAVA_HOME/bin/java -classpath "$classpath" "$class" "''${@:2}"
```

Does `$JAVA_PATH` even work for you? I made a NixOS test for this and `JAVA_PATH` does not seem to get set by the hook at all. I had to use `$JAVA_HOME/bin/java` for the test to succeed

---

Didn't know `$JAVA_PATH` doesn't get set. I'll fix that.

---

Okay, cool. So it's not just me. Here's the NixOS test I'm using, for the record:

<details>
<summary><code>nixos/tests/java.nix</code></summary>

```nix
import ./make-test-python.nix ({ pkgs, lib, ... }:
  let
    testJava = pkgs.writeText "Test.java" ''
        public class Test {
          public static void main(String[] args) {
            System.out.print(args[0]);
          }
        }
    '';

    java-binfmt-test = with pkgs; stdenv.mkDerivation {
      name = "java-binfmt-test";

      src = testJava;

      nativeBuildInputs = [ jdk ];

      unpackPhase = "cp $src Test.java";
      buildPhase = ''
        javac Test.java
        jar cvfe Test.jar Test Test.class
      '';
      installPhase = ''
        install -Dm0755 Test.class $out/share/java/Test.class
        install -Dm0755 Test.jar $out/share/java/Test.jar
      '';
    };
  in
{
  name = "java";

  nodes.machine = {
    programs.java = {
      enable = true;
      binfmt = true;
    };
  };

  testScript = ''
    machine.wait_for_unit("multi-user.target")
    assert machine.succeed("${java-binfmt-test}/share/java/Test.class shibboleth") == "shibboleth"
    assert machine.succeed("${java-binfmt-test}/share/java/Test.jar shibboleth") == "shibboleth"
  '';
})
```
</details>

---

I think `bintfmt` should require `enable` because it's more like `binfmt` is a subset of java and `enable` enables Java.

---

They don't rely on each other functionally (one adds a binfmt registration regardless of the system PATH and the other adds java to system PATH and makes sure JAVA_HOME is globally set), but that sounds fine enough to me